components/fxa-client/src/oauth.rs

```diff
@@ -314,7 +313,40 @@ impl FirefoxAccount {
         // When our keys change, we might need to re-register device capabilities with the server.
         // Ensure that this happens on the next call to ensure_capabilities.
         self.state.device_capabilities.clear();
+        Ok(())
+    }
 
+    /// Typically called during a password change flow.
+    /// Invalidates all tokens and fetches a new refresh token.
+    /// Because the old refresh token is not valid anymore, we can't do like `handle_oauth_response`
+    /// and re-create the device, so it is the responsibility of the caller to do so after we're
+    /// done.
+    ///
+    /// **💾 This method alters the persisted account state.**
+    pub fn handle_session_token_change(&mut self, session_token: &str) -> Result<()> {
+        self.state.session_token = Some(session_token.to_owned());
+        self.state.access_token_cache.clear();
```

Comment:
Nit: what do you think about avoiding mutating `self.state` until the end of this function, after we've done all of the network calls? I think it would be useful to guarantee that we don't risk putting `self.state` into an inconsistent state if execution happens to be interrupted (e.g. if we set sessionToken at the start, but get a network error partway through and end up leaving the old refresh_token behind).

It doesn't matter right now because we don't persist state changes on error, but it might matter in the future if we change how state is handled, and it could be easy to forget and come back to each of these functions and guard against such inconsistency.